File: packages/mobile/modules/tlsn-native/ios/TlsnNativeModule.swift
import ExpoModulesCore
import os.log

// The tlsn_mobile.swift file is compiled together with this module,
// making initialize(), prove(), HttpHeader, etc. available as top-level symbols.

private let logger = OSLog(subsystem: "com.tlsn.mobile", category: "TlsnNative")

public class TlsnNativeModule: Module {
    public func definition() -> ModuleDefinition {
        Name("TlsnNative")

        // Initialize the TLSN library
        Function("initialize") {
            do {
                try initialize()
            } catch {
                throw Exception(name: "TlsnError", description: "Failed to initialize: \(error)")
            }
        }

        // High-level prove function that matches the React Native interface
        AsyncFunction("prove") { (requestDict: [String: Any], optionsDict: [String: Any], promise: Promise) in
            DispatchQueue.global(qos: .userInitiated).async {
                do {
                    // Parse HTTP request
                    guard let url = requestDict["url"] as? String,
                          let method = requestDict["method"] as? String else {
                        promise.reject("InvalidRequest", "Missing url or method")
                        return
                    }

                    var headers: [HttpHeader] = []
                    if let headersDict = requestDict["headers"] as? [String: String] {
                        for (name, value) in headersDict {
                            headers.append(HttpHeader(name: name, value: value))
                        }
                    }

                    let body = requestDict["body"] as? String

                    let request = HttpRequest(
                        url: url,
                        method: method,
                        headers: headers,
                        body: body
                    )

                    // Parse prover options
                    guard let verifierUrl = optionsDict["verifierUrl"] as? String,
                          let proxyUrl = optionsDict["proxyUrl"] as? String else {
                        promise.reject("InvalidOptions", "Missing verifierUrl or proxyUrl")
                        return
                    }

                    let maxSentData = (optionsDict["maxSentData"] as? NSNumber)?.uint32Value ?? 4096
                    let maxRecvData = (optionsDict["maxRecvData"] as? NSNumber)?.uint32Value ?? 16384

                    // Parse handlers for selective disclosure
                    var handlers: [Handler] = []
                    print("[TlsnNative] optionsDict keys: \(optionsDict.keys)")
                    print("[TlsnNative] handlers raw value: \(String(describing: optionsDict["handlers"]))")

                    if let handlersArray = optionsDict["handlers"] as? [[String: Any]] {
                        print("[TlsnNative] Found \(handlersArray.count) handlers to parse")
                        for (index, handlerDict) in handlersArray.enumerated() {
                            print("[TlsnNative] Handler \(index) dict: \(handlerDict)")

                            guard let handlerTypeStr = handlerDict["handlerType"] as? String else {
                                print("[TlsnNative] Handler \(index): missing handlerType")
                                continue
                            }
                            guard let partStr = handlerDict["part"] as? String else {
                                print("[TlsnNative] Handler \(index): missing part")
                                continue
                            }
                            guard let actionStr = handlerDict["action"] as? String else {
                                print("[TlsnNative] Handler \(index): missing action")
                                continue
                            }

                            print("[TlsnNative] Handler \(index): type=\(handlerTypeStr), part=\(partStr), action=\(actionStr)")

                            // Parse handler type
                            let handlerType: HandlerType
                            switch handlerTypeStr {
                            case "Sent": handlerType = .sent
                            case "Recv": handlerType = .recv
                            default:
                                print("[TlsnNative] Handler \(index): unknown handlerType '\(handlerTypeStr)'")
                                continue
                            }

                            // Parse handler part
                            let part: HandlerPart
                            switch partStr {
                            case "StartLine": part = .startLine
                            case "Headers": part = .headers
                            case "Body": part = .body
                            case "All": part = .all
                            default:
                                print("[TlsnNative] Handler \(index): unknown part '\(partStr)'")
                                continue
                            }

                            // Parse handler action
                            let action: HandlerAction
                            switch actionStr {
                            case "Reveal": action = .reveal
                            default:
                                print("[TlsnNative] Handler \(index): unknown action '\(actionStr)'")
                                continue
                            }

                            // Parse optional params
                            var params: HandlerParams? = nil
                            if let paramsDict = handlerDict["params"] as? [String: Any] {
                                let key = paramsDict["key"] as? String
                                let contentType = paramsDict["contentType"] as? String
                                let path = paramsDict["path"] as? String
                                params = HandlerParams(key: key, contentType: contentType, path: path)
                                print("[TlsnNative] Handler \(index) params: key=\(String(describing: key)), contentType=\(String(describing: contentType)), path=\(String(describing: path))")
                            }

                            handlers.append(Handler(handlerType: handlerType, part: part, action: action, params: params))
                            print("[TlsnNative] Handler \(index) successfully parsed")
                        }
                    } else {
                        print("[TlsnNative] Could not parse handlers array - checking raw type")
                        if let rawHandlers = optionsDict["handlers"] {
                            print("[TlsnNative] handlers type: \(type(of: rawHandlers))")
                        } else {
                            print("[TlsnNative] handlers key not found in optionsDict")
                        }
                    }

                    NSLog("[TlsnNative] Final handlers count: %d", handlers.count)
                    os_log("[TlsnNative] Handlers to Rust: %d", log: logger, type: .info, handlers.count)

                    let options = ProverOptions(
                        verifierUrl: verifierUrl,
                        proxyUrl: proxyUrl,
                        maxSentData: maxSentData,
                        maxRecvData: maxRecvData,
                        handlers: handlers
                    )

                    // Call the prove function
                    let result = try prove(request: request, options: options)

                    // Convert response headers to dictionary
                    var responseHeaders: [[String: String]] = []
                    for header in result.response.headers {
                        responseHeaders.append(["name": header.name, "value": header.value])
                    }

                    // Parse response body as JSON if possible
                    var bodyJson: Any = result.response.body
                    if let jsonData = result.response.body.data(using: .utf8),
                       let json = try? JSONSerialization.jsonObject(with: jsonData) {
                        bodyJson = json
                    }

                    // Log transcript info
                    NSLog("[TlsnNative] Proof complete! Sent: %d bytes, Recv: %d bytes",
                          result.transcript.sent.count, result.transcript.recv.count)

                    // Build result dictionary
                    let resultDict: [String: Any] = [
                        "status": Int(result.response.status),
                        "headers": responseHeaders,
                        "body": bodyJson,
                        "transcript": [
                            "sentLength": result.transcript.sent.count,
                            "recvLength": result.transcript.recv.count
                        ],
                        "debug": [
                            "handlersPassedToRust": handlers.count,
                            "handlersReceivedByRust": result.handlersReceived
                        ]
                    ]

                    promise.resolve(resultDict)

                } catch let error as TlsnError {
                    promise.reject("TlsnError", "\(error)")
                } catch {
                    promise.reject("UnknownError", error.localizedDescription)
                }
            }
        }

        // Check if native module is available
        Function("isAvailable") { () -> Bool in
            return true
        }
    }
}

File: packages/mobile/modules/tlsn-native/android/src/main/java/expo/modules/tlsnnative/TlsnNativeModule.kt
package expo.modules.tlsnnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import uniffi.tlsn_mobile.HttpHeader
import uniffi.tlsn_mobile.HttpRequest
import uniffi.tlsn_mobile.ProverOptions
import uniffi.tlsn_mobile.Handler
import uniffi.tlsn_mobile.HandlerType
import uniffi.tlsn_mobile.HandlerPart
import uniffi.tlsn_mobile.HandlerAction
import uniffi.tlsn_mobile.HandlerParams
import uniffi.tlsn_mobile.initialize as rustInitialize
import uniffi.tlsn_mobile.prove as rustProve
import org.json.JSONObject
import org.json.JSONArray

/** Convert org.json types to native Map/List that Expo can bridge to JS. */
private fun jsonToNative(value: Any?): Any? = when (value) {
    is JSONObject -> {
        val map = mutableMapOf<String, Any?>()
        for (key in value.keys()) {
            map[key] = jsonToNative(value.get(key))
        }
        map
    }
    is JSONArray -> {
        val list = mutableListOf<Any?>()
        for (i in 0 until value.length()) {
            list.add(jsonToNative(value.get(i)))
        }
        list
    }
    JSONObject.NULL -> null
    else -> value
}

class TlsnNativeModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("TlsnNative")

        Function("initialize") {
            rustInitialize()
        }

        // Accept JSON strings to avoid Expo Kotlin bridge type conversion issues
        // with nested objects. The JS side JSON.stringifies before calling.
        AsyncFunction("prove") { requestJson: String, optionsJson: String, promise: Promise ->
            Thread {
                try {
                    val requestObj = JSONObject(requestJson)
                    val optionsObj = JSONObject(optionsJson)

                    // Parse HTTP request
                    val url = requestObj.optString("url", "")
                        .ifEmpty { return@Thread promise.reject("InvalidRequest", "Missing url", null) }
                    val method = requestObj.optString("method", "")
                        .ifEmpty { return@Thread promise.reject("InvalidRequest", "Missing method", null) }

                    val headers = mutableListOf<HttpHeader>()
                    val headersJsonObj = requestObj.optJSONObject("headers")
                    if (headersJsonObj != null) {
                        for (name in headersJsonObj.keys()) {
                            headers.add(HttpHeader(name, headersJsonObj.getString(name)))
                        }
                    }

                    val body: String? = if (requestObj.has("body") && !requestObj.isNull("body")) {
                        requestObj.getString("body")
                    } else null

                    val request = HttpRequest(
                        url = url,
                        method = method,
                        headers = headers,
                        body = body
                    )

                    // Parse prover options
                    val verifierUrl = optionsObj.optString("verifierUrl", "")
                        .ifEmpty { return@Thread promise.reject("InvalidOptions", "Missing verifierUrl", null) }

                    val maxSentData = optionsObj.optInt("maxSentData", 4096).toUInt()
                    val maxRecvData = optionsObj.optInt("maxRecvData", 16384).toUInt()

                    // Parse handlers for selective disclosure
                    val handlers = mutableListOf<Handler>()
                    val handlersArray = optionsObj.optJSONArray("handlers")

                    if (handlersArray != null) {
                        for (index in 0 until handlersArray.length()) {
                            val handlerObj = handlersArray.getJSONObject(index)
                            val handlerTypeStr = handlerObj.optString("handlerType", "")
                            val partStr = handlerObj.optString("part", "")
                            val actionStr = handlerObj.optString("action", "")

                            if (handlerTypeStr.isEmpty() || partStr.isEmpty() || actionStr.isEmpty()) continue

                            val handlerType = when (handlerTypeStr) {
                                "Sent" -> HandlerType.SENT
                                "Recv" -> HandlerType.RECV
                                else -> continue
                            }

                            val part = when (partStr) {
                                "StartLine" -> HandlerPart.START_LINE
                                "Protocol" -> HandlerPart.PROTOCOL
                                "Method" -> HandlerPart.METHOD
                                "RequestTarget" -> HandlerPart.REQUEST_TARGET
                                "StatusCode" -> HandlerPart.STATUS_CODE
                                "Headers" -> HandlerPart.HEADERS
                                "Body" -> HandlerPart.BODY
                                "All" -> HandlerPart.ALL
                                else -> continue
                            }

                            val action = when (actionStr) {
                                "Reveal" -> HandlerAction.REVEAL
                                "Pedersen" -> HandlerAction.PEDERSEN
                                else -> continue
                            }

                            val paramsObj = handlerObj.optJSONObject("params")
                            val params = if (paramsObj != null) {
                                HandlerParams(
                                    key = if (paramsObj.has("key")) paramsObj.optString("key") else null,
                                    hideKey = if (paramsObj.has("hideKey")) paramsObj.optBoolean("hideKey") else null,
                                    hideValue = if (paramsObj.has("hideValue")) paramsObj.optBoolean("hideValue") else null,
                                    contentType = if (paramsObj.has("contentType")) paramsObj.optString("contentType") else null,
                                    path = if (paramsObj.has("path")) paramsObj.optString("path") else null,
                                    regex = if (paramsObj.has("regex")) paramsObj.optString("regex") else null,
                                    flags = if (paramsObj.has("flags")) paramsObj.optString("flags") else null
                                )
                            } else {
                                null
                            }

                            handlers.add(Handler(handlerType, part, action, params))
                            android.util.Log.i("TlsnNative", "Handler $index: type=$handlerTypeStr, part=$partStr, action=$actionStr")
                        }
                    }

                    android.util.Log.i("TlsnNative", "Final handlers count: ${handlers.size}")

                    val options = ProverOptions(
                        verifierUrl = verifierUrl,
                        maxSentData = maxSentData,
                        maxRecvData = maxRecvData,
                        handlers = handlers
                    )

                    // Call the prove function
                    val result = rustProve(request, options)

                    // Convert response headers
                    val responseHeaders = result.response.headers.map { header ->
                        mapOf("name" to header.name, "value" to header.value)
                    }

                    // Parse response body as JSON, converting to native Map/List for Expo bridge
                    val bodyJson: Any = try {
                        jsonToNative(JSONObject(result.response.body))!!
                    } catch (_: Exception) {
                        try {
                            jsonToNative(JSONArray(result.response.body))!!
                        } catch (_: Exception) {
                            result.response.body
                        }
                    }

                    android.util.Log.i("TlsnNative",
                        "Proof complete! Sent: ${result.transcript.sent.size} bytes, Recv: ${result.transcript.recv.size} bytes")

                    // Build result dictionary
                    val resultMap = mapOf(
                        "status" to result.response.status.toInt(),
                        "headers" to responseHeaders,
                        "body" to bodyJson,
                        "transcript" to mapOf(
                            "sentLength" to result.transcript.sent.size,
                            "recvLength" to result.transcript.recv.size
                        ),
                        "debug" to mapOf(
                            "handlersPassedToRust" to handlers.size,
                            "handlersReceivedByRust" to result.handlersReceived.toInt()
                        )
                    )

                    promise.resolve(resultMap)

                } catch (e: uniffi.tlsn_mobile.TlsnException) {
                    promise.reject("TlsnError", "$e", null)
                } catch (e: Exception) {
                    promise.reject("UnknownError", e.localizedMessage, null)
                }
            }.start()
        }

        Function("isAvailable") {
            true
        }
    }
}
